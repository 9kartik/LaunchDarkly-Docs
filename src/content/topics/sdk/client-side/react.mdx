---
path: '/sdk/client-side/react'
title: 'React SDK Reference'
description: 'This topic explains all of the methods available in the client-side React SDK, with detail about how they work.'
published: true
---

This reference guide documents all of the methods available in our client-side React SDK, and explains in detail how these methods work. If you want to dig even deeper, our SDKs are open source â€” head to the [React SDK code](https://github.com/launchdarkly/react-client-sdk) in our JavaScript SDK GitHub repository to look under the hood. The online [API docs](https://launchdarkly.github.io/react-client-sdk/) contain the programmatic definitions of every type and method. Additionally you can clone and run a [sample application](/sdk/client-side/react#example-app) using this SDK.

<Callout intent="info">
  <CalloutTitle>React SDK is based on the JavaScript SDK</CalloutTitle>
  <CalloutDescription>

The React SDK builds on top of LaunchDarkly's JavaScript SDK to provide a better integration for use in React
applications. As a result, much of the JavaScript SDK functionality is also available for use by the React SDK. For
a complete client-side JavaScript SDK reference, read
[JavaScript SDK reference](/sdk/client-side/javascript).

  </CalloutDescription>
</Callout>
<Callout intent="alert">
  <CalloutTitle>Requires React 16.3.0 or later</CalloutTitle>
  <CalloutDescription>

This SDK uses the Context API, which requires React version 16.3.0 or later. Please note that you must use a newer
version if you wish to additionally use the Hooks API. To learn more, read [Hooks](#hooks).

  </CalloutDescription>
</Callout>

## Getting started

Building on top of our [Getting Started guide](/home/getting-started) guide, the following steps will get you started with using the LaunchDarkly SDK in your React code.

You can use either `npm` or `yarn` to install the React SDK:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'Installing with npm', value: 'js', },
    { label: 'Installing with yarn', value: 'yarn', },
  ]
}>
<CodeTabItem value="js">

```bash
npm i --save launchdarkly-react-client-sdk
```

</CodeTabItem>
<CodeTabItem value="yarn">

```bash
yarn add launchdarkly-react-client-sdk
```

</CodeTabItem>
</CodeTabs>

Once the dependency is installed, you'll want to initialize the React SDK. To do this, you need your environment's client-side ID (available on your [account settings page](https://app.launchdarkly.com/settings#/projects)). Client-side IDs are not secret-- they can be safely exposed in your client-side code.

There are two ways you can initialize the React SDK. The first is using the `asyncWithLDProvider` function. The second is using the `withLDProvider` function. Both rely on React's `Context` API to make it easy for you to access your flags from any level of your component hierarchy. Both functions accept a [`ProviderConfig`](/sdk/client-side/react#configuring-the-react-sdk) object used to configure the React SDK. Continue reading below for more information about these functions.

## `asyncWithLDProvider`

The `asyncWithLDProvider` function initializes the React SDK and returns a provider which is a React component. It is an async function.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
import { asyncWithLDProvider } from 'launchdarkly-react-client-sdk';

(async () => {
  const LDProvider = await asyncWithLDProvider({
    clientSideID: 'your-client-side-id',
    user: {
      "key": "aa0ceb",
      "name": "Grace Hopper",
      "email": "gracehopper@example.com"
 	  },
	  options: { /* ... */ }
  });

  render(
    <LDProvider>
      <YourApp />
    </LDProvider>,
    document.getElementById('reactDiv'),
  );
})();
```

</CodeTabItem>
</CodeTabs>

Once initialization is complete, your flags and the ldClient will be available at the start of your React app lifecycle. This will ensure your app does not flicker due to flag changes at startup time.

<Callout intent="alert">
  <CalloutTitle>Delayed rendering</CalloutTitle>
  <CalloutDescription>

The asynchronous nature of the `asyncWithLDProvider` function means that the rendering of your React app will be
delayed until initialization is completed. Typically this should take only a few milliseconds but can take up to 200
milliseconds. Alternatively you can use the
[`withLDProvider`](/sdk/client-side/react#withldprovider) function if you prefer
to render your app first and then process flag updates once rendering is complete.

  </CalloutDescription>
</Callout>
<Callout intent="alert">
  <CalloutTitle>Requires React 16.8.0 or later</CalloutTitle>
  <CalloutDescription>

The `asyncWithLDProvider` function uses the Hooks API, which requires React version 16.8.0 or later.

  </CalloutDescription>
</Callout>

## `withLDProvider`

The `withLDProvider` function initializes the React SDK and wraps your root component in a `Context.Provider`:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
import { withLDProvider } from 'launchdarkly-react-client-sdk';

export default withLDProvider({
  clientSideID: 'your-client-side-id',
  user: {
    "key": "aa0ceb",
    "name": "Grace Hopper",
    "email": "gracehopper@example.com"
  },
  options: { /* ... */ }
})(YourApp);
```

</CodeTabItem>
</CodeTabs>

Customizing your client can be done via the `options` field. See the [JavaScript SDK reference](/sdk/client-side/javascript#customizing-your-client) for all the available options.

Note that the React SDK automatically subscribes to flag change events; this is unlike the [JavaScript SDK Reference](/sdk/client-side/javascript#subscribing-to-feature-flag-changes) where customers need to opt in to event listening. Consequently, in the React SDK, streaming mode is enabled by default. To disable streaming mode, specify a `streaming: false` attribute in your `options` object.

## `withLDConsumer`

The return value of `withLDConsumer` is a wrapper function that takes your component and returns a React component injected with `flags` & `ldClient` as props.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
import { withLDConsumer } from 'launchdarkly-react-client-sdk';

const Home = ({ flags, ldClient /*, ...otherProps */ }) => {
  // You can call any of the methods from the JavaScript SDK
  // ldClient.identify({...})

  return flags.devTestFlag ? <div>Flag on</div> : <div>Flag off</div>;
};

export default withLDConsumer()(Home);
```

</CodeTabItem>
</CodeTabs>

## Configuring the React SDK

The [`ProviderConfig`](https://launchdarkly.github.io/react-client-sdk/interfaces/providerconfig.html) object is used to configure the React SDK. The only mandatory property is the client-side id. All other properties are optional.

- `clientSideId` - This is your project and environment specific client-side id. This property is mandatory.

- `user` - A LaunchDarkly [user](/sdk/client-side/javascript#users) object. This property is optional.

- `options` - LaunchDarkly JavaScript SDK initialization [options](/sdk/client-side/javascript#customizing-your-client). This property is optional.

- `reactOptions` - This property is specific to the React SDK. You can use this option to [disable automatic camel casing](/sdk/client-side/react#flag-keys) of flag keys when using the React SDK. This property is optional.

- `flags` - If specified, the React SDK will only request and subscribe for updates to these flags. The React SDK will request for and subscribe to all flags when this property is unspecified. This property is optional.

## Hooks

The React SDK offers two custom hooks which you can use as an alternative to `withLDConsumer`: `useFlags` and `useLDClient`.

<Callout intent="alert">
  <CalloutTitle>Requires React 16.8.0 or later</CalloutTitle>
  <CalloutDescription>

`useFlags` and `useLDClient` use the Hooks API, which requires React version 16.8.0 or later.

  </CalloutDescription>
</Callout>

`useFlags` is a custom hook which returns all feature flags. It uses the useContext primitive to access the LaunchDarkly context set up by `asyncWithLDProvider` or `withLDProvider`. You will still need to use the `asyncWithLDProvider` or the `withLDProvider` higher-order component at the root of your application to initialize the React SDK and populate the context with `LDClient` and your flags.

`useLDClient` is the second custom hook which returns the underlying LaunchDarkly JavaScript SDK client object. Like the `useFlags` custom hook, `useLDClient` also uses the useContext primitive to access the LaunchDarkly context set up by `asyncWithLDProvider` or `withLDProvider`. You will still need to use the `asyncWithLDProvider` or the `withLDProvider` higher-order component to initialize the React SDK to use this custom hook.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
import React from 'react';
import { useFlags, useLDClient } from 'launchdarkly-react-client-sdk';

const HooksDemo = () => {
  const { devTestFlag } = useFlags();
  const ldClient = useLDClient();

  const onLoginSuccessful = () => ldClient.identify({ key: 'aa0ceb' });

  return (
    <div>{devTestFlag ? 'Flag on' : 'Flag off'}</div>
  );
};

export default HooksDemo;
```

</CodeTabItem>
</CodeTabs>

## Users

All user-related functionality provided by the JavaScript SDK is equally accessible in the React SDK. Refer to the [users](/sdk/client-side/javascript#users), [private user attributes](/sdk/client-side/javascript#private-user-attributes), and [anonymous users](/sdk/client-side/javascript#anonymous-users) sections of the JavaScript SDK reference guide for more information.

## Flag keys

LaunchDarkly primarily identifies feature flags by a key which must only contain alphanumeric characters, dots (`.`), underscores (`_`), and dashes (`-`). These keys are used across all of LaunchDarkly's APIs as well as in the SDKs to identify flags.

However, keys in this format are slightly inconvenient to access in JavaScript and React as they cannot be accessed with the dot notation. As a result, the React SDK automatically camel-cases all flag keys. A flag with key `dev-flag-test` would be accessible as `flags.devFlagTest`.

Automatically camel-casing flag keys poses some problems, though:

1. It is possible to induce a key collision if there are multiple LaunchDarkly flag keys which resolve to the same camel-cased key. For example, `dev-flag-test` and `dev.flag.test` would be considered unique keys on LaunchDarkly yet the React SDK would consider them to have the same camel-cased keys.
2. Our [Git code references](/integrations/git-code-references) tool expects your source code to reference the exact key -- not a camel-cased equivalent -- and as a result, the tool does not support detecting references of keys that were automatically camel-cased.
3. As the camel-casing functionality is implemented in the React SDK instead of in the underlying JavaScript SDK, the underlying `ldClient` object and functionality provided by the JavaScript SDK reflect flag keys in their original format. Only React-specific contexts such as your injected props are camel-cased.

To disable the SDK's automatic camel-casing feature, you can provide `asyncWithLDProvider` or `withLDProvider` with the `reactOptions.useCamelCaseFlagKeys` property as shown in the following example. This is supported in versions 2.13.0 and later of the React SDK.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
export default withLDProvider({
  clientSideID: 'your-client-side-id',
  user: {
    "key": "aa0ceb",
    "name": "Grace Hopper",
    "email": "gracehopper@example.com"
  },
  reactOptions: {
    useCamelCaseFlagKeys: false
  }
})(App);
```

</CodeTabItem>
</CodeTabs>

With this configuration option in place, you would now access your `dev-flag-test` flag using bracket notation -- for example, `flag['dev-flag-test']`.

## Example App

Check the [example](https://github.com/launchdarkly/react-client-sdk/tree/master/examples/hoc) for a fully working single page app with react and react-router. You'll need to enter your `clientSideId` in the [client root app file](https://github.com/launchdarkly/react-client-sdk/blob/master/examples/hoc/src/universal/app.js) and create a flag with key `dev-test-flag` in your dashboard before running the example.
