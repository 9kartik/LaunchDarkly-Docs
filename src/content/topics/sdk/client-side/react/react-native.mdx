---
path: '/sdk/client-side/react/react-native'
title: 'React Native SDK reference'
description: 'This topic explains all of the methods available in the client-side React Native, with detail about how they work.'
published: true
---

<Callout intent="alert">
  <CalloutTitle>SDK version compatibility</CalloutTitle>
<CalloutDescription>

The latest version of the SDK is compatible with Xcode 11.4 and React Native 0.63. For compatibility with earlier React Native versions, please refer to older SDK versions.

</CalloutDescription>
</Callout>

This reference guide documents all of the methods available in our React Native SDK, and explains in detail how these methods work. If you want to dig even deeper, our SDKs are open source. To learn more, read [React Native SDK GitHub repository](https://github.com/launchdarkly/react-native-client-sdk). The online [API docs](https://launchdarkly.github.io/react-native-client-sdk/) contain the programmatic definitions of every type and method. Additionally you can clone and run a [sample application](https://github.com/launchdarkly/hello-react-native) using this SDK.

To learn more about the specific configuration options that are available in this SDK, read the [SDK's API docs](https://launchdarkly.github.io/react-native-client-sdk/index.html#ldclientconfig).

## Getting Started

<Callout intent="warning">
  <CalloutTitle>Expo usage</CalloutTitle>
<CalloutDescription>

The LaunchDarkly React Native Client SDK does not work with Expo because the SDK uses native modules. Please consider using [ExpoKit](https://docs.expo.io/expokit/expokit/).

</CalloutDescription>
</Callout>

Building on top of our [Getting Started guide](/home/getting-started), the following steps will get you started with using the LaunchDarkly SDK in your React Native code.

The first step is to install the LaunchDarkly SDK as a dependency. Add the LaunchDarkly npm dependency to your project and link it to your React Native project.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'Installing with npm', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```bash
npm install --save launchdarkly-react-native-client-sdk
cd ios/
pod install
```

</CodeTabItem>
</CodeTabs>

You will need a `Podfile` in your `ios/` directory. Once you've written your `Podfile`, run `pod install` in the `ios/` directory. This SDK currently requires Xcode 10.2.1 (it is compatible with Xcode 10.3) and is compiled against Swift 5.

<Callout intent="info">
  <CalloutTitle><code>use_frameworks!</code> and undefined symbols errors</CalloutTitle>
  <CalloutDescription>As of v2.1.0 <code>use_frameworks!</code> is not necessary in the Podfile. If you encounter an undefined symbols error when building without <code>use_frameworks!</code> in your <code>Podfile</code> you will need to create a new empty Swift file and accept the prompt to create a matching Objective-C bridging header. If you are building for multiple platforms, add those as targets for the bridging header. If you're still having trouble resolving the undefined symbols error, clean and rebuild and try again.</CalloutDescription>
</Callout>

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
platform :ios, '10.0'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

target 'YOUR_MODULE_NAME' do
  # Pods for ReactNativeRestwrapper
  pod 'FBLazyVector', :path => "../node_modules/react-native/Libraries/FBLazyVector"
  pod 'FBReactNativeSpec', :path => "../node_modules/react-native/Libraries/FBReactNativeSpec"
  pod 'RCTRequired', :path => "../node_modules/react-native/Libraries/RCTRequired"
  pod 'RCTTypeSafety', :path => "../node_modules/react-native/Libraries/TypeSafety"
  pod 'React', :path => '../node_modules/react-native'
  pod 'React-Core', :path => '../node_modules/react-native'
  pod 'React-CoreModules', :path => '../node_modules/react-native/React/CoreModules'
  pod 'React-Core/DevSupport', :path => '../node_modules/react-native'
  pod 'React-RCTActionSheet', :path => '../node_modules/react-native/Libraries/ActionSheetIOS'
  pod 'React-RCTAnimation', :path => '../node_modules/react-native/Libraries/NativeAnimation'
  pod 'React-RCTBlob', :path => '../node_modules/react-native/Libraries/Blob'
  pod 'React-RCTImage', :path => '../node_modules/react-native/Libraries/Image'
  pod 'React-RCTLinking', :path => '../node_modules/react-native/Libraries/LinkingIOS'
  pod 'React-RCTNetwork', :path => '../node_modules/react-native/Libraries/Network'
  pod 'React-RCTSettings', :path => '../node_modules/react-native/Libraries/Settings'
  pod 'React-RCTText', :path => '../node_modules/react-native/Libraries/Text'
  pod 'React-RCTVibration', :path => '../node_modules/react-native/Libraries/Vibration'
  pod 'React-Core/RCTWebSocket', :path => '../node_modules/react-native'

  pod 'React-cxxreact', :path => '../node_modules/react-native/ReactCommon/cxxreact'
  pod 'React-jsi', :path => '../node_modules/react-native/ReactCommon/jsi'
  pod 'React-jsiexecutor', :path => '../node_modules/react-native/ReactCommon/jsiexecutor'
  pod 'React-jsinspector', :path => '../node_modules/react-native/ReactCommon/jsinspector'
  pod 'React-callinvoker', :path => "../node_modules/react-native/ReactCommon/callinvoker"
  pod 'ReactCommon/turbomodule/core', :path => "../node_modules/react-native/ReactCommon"
  pod 'Yoga', :path => '../node_modules/react-native/ReactCommon/yoga'

  pod 'DoubleConversion', :podspec => '../node_modules/react-native/third-party-podspecs/DoubleConversion.podspec'
  pod 'glog', :podspec => '../node_modules/react-native/third-party-podspecs/glog.podspec'
  pod 'Folly', :podspec => '../node_modules/react-native/third-party-podspecs/Folly.podspec'

  #launchdarkly-react-native-client-sdk dependencies
  pod 'LaunchDarkly', '4.5.0'

  use_native_modules!
end
```

</CodeTabItem>
</CodeTabs>

Once the SDK is downloaded and saved in `package.json` and the other install steps are complete, you'll need to import it into your project:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
import LDClient from 'launchdarkly-react-native-client-sdk';
```

</CodeTabItem>
</CodeTabs>

Once the SDK is installed and imported, you'll want to create a single, shared instance of LDClient, passing in the client and user configuration objects. To create a client instance, you need your environment's mobile key (available on your [account settings page](https://app.launchdarkly.com/settings#/projects)). Mobile keys are not secret. You can expose them in your client-side code with no risk.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let client = new LDClient();
let clientConfig = { "mobileKey": "YOUR_MOBILE_KEY" };
let userConfig = { "key": "user_key" };

await client.configure(clientConfig, userConfig);
```

</CodeTabItem>
</CodeTabs>

Lastly, when your application is about to terminate, shut down LDClient. This ensures that the client releases any resources it is using, and that any pending analytics events are delivered to LaunchDarkly. If your application quits without this shutdown step, you may not see your requests and users on the dashboard, because they are derived from analytics events. This is something you only need to do once.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
client.close();
```

</CodeTabItem>
</CodeTabs>

<Callout intent="info">
  <CalloutTitle>Mobile keys</CalloutTitle>
<CalloutDescription>

Be sure to use a mobile key from your [Environments](https://app.launchdarkly.com/settings#/environments) page.

Never embed a server-side SDK key into a mobile application.

</CalloutDescription>
</Callout>

## Client Configuration Options

You can also pass custom parameters to the client by creating a custom configuration object:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let clientConfig =
  { "mobileKey": "YOUR_MOBILE_KEY",
    "connectionTimeoutMillis": 3000
  };
await client.configure(clientConfig, userConfig);
```

</CodeTabItem>
</CodeTabs>

The `clientConfig` object can include a variety of options. `mobileKey` is a required property, but all other properties are optional. To learn more about the specific configuration options that are available in this SDK, read the [SDK's API docs](https://launchdarkly.github.io/react-native-client-sdk/index.html#ldclientconfig).

## User Configuration Options

Here is a complete user configuration object utilizing all available fields:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let userConfig =
  { key: '2fbfa269b78e24c8e967f482b89c939b',
    name: 'grace_hopper',
    firstName: 'Grace',
    lastName: 'Hopper',
    email: 'gracehopper@example.net',
    anonymous: false,
    privateAttributeNames: ['name'],
    country: 'USA',
    custom: {'groups': 'users'}
  };
```

</CodeTabItem>
</CodeTabs>

Let's walk through this code. The first key in the object is the user's key. In this case we've used the value `2fbfa269b78e24c8e967f482b89c939b`. The user key is the only mandatory user attribute. The key should also uniquely identify each user. You can use any value such as a primary key, an e-mail address, or a hash, as long as the same user always has the same key. We recommend using a hash if possible.

All of the other keys (like firstName, email, and the custom attributes) are optional. The attributes you specify will automatically appear on the LaunchDarkly application dashboard, meaning that you can start segmenting and targeting users with these attributes.

In addition to built-in attributes like names and e-mail addresses, you can pass us any of your own user data by passing custom attributes, like the groups attribute in the example above. Custom attributes are one of the most powerful features of LaunchDarkly. They let you target users according to any data that you want to send to us, including organizations, groups, andaccount plans. Anything you pass to us becomes available instantly on your dashboard.

Private attributes allow you to specify keys you do not want to be sent to the LaunchDarkly servers. Private user attributes can be used for targeting purposes, but are removed from the user data sent back to LaunchDarkly. When this user is sent back to LaunchDarkly, the `name` attribute will be removed.

The `anonymous` key allows you to distinguish logged out or unregistered users while still using LaunchDarkly. Anonymous users work just like regular users, except that they won't appear on your Users page in LaunchDarkly. This keeps anonymous users from polluting your Users page.

You can't search for anonymous users on your Features page, and you can't search or autocomplete by anonymous user keys. If you use the same user key for every anonymous user, you also can't use percentage rollouts or Experimentation with anonymous users.

We recommend using the same user key for every initialization and then replacing that with the actual user key when you know who the user is. This way LaunchDarkly counts the initialization user key only once against your MAU, instead of every time you initialize.

## Variations
The variation method determines whether a flag is enabled or not for a specific user. In React Native, there is a variation method for each type (e.g. `boolVariation`,` stringVariation`):

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let jsonResult = await client.jsonVariation('MY_JSON_FLAG_KEY', {});
```

</CodeTabItem>
</CodeTabs>

Variation calls take the feature flag key and a fallback value.

The fallback value will only be returned if an error is encountered. For example, the default value returns if the feature flag key doesn't exist or the user doesn't have a key specified.

The variation call will automatically create a user in LaunchDarkly if a user with that user key doesn't exist already. There's no need to create users ahead of their initial use, but if you choose to do so, you can use the `identify` method to do so.

<Callout intent="info">
  <CalloutTitle>Handling flag values on initial application launch</CalloutTitle>
 <CalloutDescription>
When LDClient is initialized for the first time at app launch, users receive feature flag fallback values until an initial connection to LaunchDarkly is completed for the first time. Read the section above for various ways to initialize the client.</CalloutDescription>
</Callout>

<Callout intent="alert">
  <CalloutTitle>Making feature flags available to this SDK</CalloutTitle>
  <CalloutDescription>
You must make feature flags available to client-side SDKs before the flags can be evaluated. If you evaluate a feature flag that is not available, you will receive the default value for that flag. If you want a project's flags to be made available to this SDK by default, check the **SDKs using Mobile Key** checkbox in your project [Settings](https://app.launchdarkly.com/settings#/projects).
  </CalloutDescription>
</Callout>

## VariationDetail

<Callout intent="info">
  <CalloutTitle>Availability</CalloutTitle>
  <CalloutDescription>Since v2.1.0</CalloutDescription>
</Callout>

The `variationDetail` methods work the same as `variation`, but also provide additional `reason` information about how a flag value was calculated (for instance, if the user matched a specific rule). You can examine the `reason` data programmatically; you can also view it with data export, if you are capturing detailed analytics events for this flag.

For more information, see [Evaluation reasons](/sdk/concepts/evaluation-reasons).

## All Flags

<Callout intent="alert">
  <CalloutTitle>Creating users</CalloutTitle>
<CalloutDescription>

Unlike `variation` and `identify` calls, `allFlags` does not send events to LaunchDarkly. When you use `allFlags`, users are not created or updated in the LaunchDarkly dashboard.

</CalloutDescription>
</Callout>

The `allFlags` method returns an object containing flag keys and values. You may also register a listener that will fire when any flag in your environment changes.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let allFlagsResult = client.allFlags();
allFlagsResult.then(values => {
  console.log(values);
});

// All Flags Listener
let listenerId = "ARBITRARY_VALUE"
let listener = value => Alert.alert('All Flags Listener Callback', value);
client.registerAllFlagsListener(listenerId, listener);
```

</CodeTabItem>
</CodeTabs>

## Track

The `track` method allows you to record actions your users take in your app. In the LaunchDarkly dashboard, you can tie these events to goals in A/B tests. You can also attach custom JSON data to your event by passing an extra parameter to `track`. Here's a simple example:

<Callout intent="info">
  <CalloutTitle>Availability</CalloutTitle>
  <CalloutDescription>Since v2.1.0</CalloutDescription>
</Callout>

You can now optionally add a `metricValue` parameter to the `track` method if you are using the latest version of Experimentation.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
client.track('MY_GOAL_FLAG_KEY', false);
client.track('MY_GOAL_FLAG_KEY_WITH_DATA', {'some_data':'value'});
```

</CodeTabItem>
</CodeTabs>

## Offline Mode
In some situations, you might want to stop making remote calls to LaunchDarkly and switch to the last known values for your feature flags. Offline mode lets you do this easily.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let offlineResult = client.setOffline(); //true or false
let offlineStatus = client.isOffline(); //true or false
let onlineResult = client.setOnline(); //true or false
```

</CodeTabItem>
</CodeTabs>

<Callout intent="info">
  <CalloutTitle>Lack of network connectivity</CalloutTitle>
<CalloutDescription>

If a user's device is in airplane/flight mode or if they are not connected to a network, LaunchDarkly will use the latest stored flag settings in memory. If there are no previously stored flag settings, then the fallback values will be used.

</CalloutDescription>
</Callout>

## Flush

Internally, the LaunchDarkly SDK keeps an event buffer for `track` calls. These are flushed periodically in a background thread. In some situations (for example, if you're testing out the SDK in a simulator), you may want to manually call `flush` to process events immediately.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
client.flush();
```

</CodeTabItem>
</CodeTabs>

## Identify

If your app is used by multiple users on a single device, you may want to change users and have separate flag settings for each user. To achieve this, the SDK will store the last 5 user contexts on a single device, with support for switching between different user contexts.

You can use the `identify` method to switch user contexts:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let user = {'key': 'minimal_user'};
await client.identify(user); //promise resolves to nil/null
Alert.alert('identify', 'success');
```

</CodeTabItem>
</CodeTabs>

The `identify` call will load any saved flag values for the new user and immediately trigger an update of the latest flags from LaunchDarkly.

## Real-time Updates

LaunchDarkly manages all flags for a user context in real-time by updating flags based on a real-time event stream. When a flag is modified from the LaunchDarkly dashboard, the flag values for the current user will update almost immediately.

To accomplish real-time updates, LaunchDarkly broadcasts an event stream that is listened to by the React Native SDK. Whenever an event is performed on the dashboard, the React Native SDK is notified of the updated flag settings in real-time.

To perform real-time updates in your app, your app will need to register listeners for each flag you'd like to watch:


<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
if (this.state.listeners.hasOwnProperty(key)) {
  return;
}
let listener = value => Alert.alert('Listener Callback', value);
client.registerFeatureFlagListener('MY_LISTEN_FLAG_KEY', listener);
this.setState({listeners: {...this.state.listeners, ...{['MY_LISTEN_FLAG_KEY']: listener}}});
```

</CodeTabItem>
</CodeTabs>

Similarly you can unregister listeners to disable them:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
this.state.ldClient.unregisterFeatureFlagListener('MY_LISTEN_FLAG_KEY', this.state.listeners['MY_LISTEN_FLAG_KEY']);
let {['MY_LISTEN_FLAG_KEY']: omit, ...newListeners} = this.state.listeners;
this.setState({listeners: newListeners});
```

</CodeTabItem>
</CodeTabs>

## Background Fetch
The React Native SDK defaults to the background fetch behavior for the platform it's running on.

## Monitoring SDK Status
The React Native SDK exposes some of its internal status through the Connection Status API to allow your application to monitor the SDK's status. This is provided primarily as a mechanism for the application to determine how recently the internal flag cache has been updated with the most recent values, as well as diagnosing potential reasons for the flag cache to be out of date. Please see the documentation for iOS or Android Connection Status for detailed explanation of the information available on that platform.

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let connectionInformation = client.getConnectionInformation();
console.log("Connection Information: " + connectionInformation);
```

</CodeTabItem>
</CodeTabs>

## Data collection

To learn more about data collection within this SDK and implications on submissions to the Apple App Store, read [Apple App Store data collection policy](/sdk/concepts/apple-app-store).
