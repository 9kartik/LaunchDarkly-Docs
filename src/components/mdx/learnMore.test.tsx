import React from 'react'
import { render } from '@testing-library/react'
import LearnMore from './learnMore'

describe('learn more', () => {
  test('renders correctly with a single link', () => {
    const { asFragment } = render(
      <LearnMore>
        <LearnMore.Link to="/foobar">Foobar</LearnMore.Link>
      </LearnMore>,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders correctly with multiple links', () => {
    const { asFragment } = render(
      <LearnMore>
        <LearnMore.Link to="/foobar">a</LearnMore.Link>
        <LearnMore.Link to="/foobar">b</LearnMore.Link>
        <LearnMore.Link to="/foobar">c</LearnMore.Link>
      </LearnMore>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
