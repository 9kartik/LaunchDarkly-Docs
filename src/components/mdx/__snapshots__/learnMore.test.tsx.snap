// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`learn more renders correctly with a single link 1`] = `
<DocumentFragment>
  .emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  box-sizing: border-box;
  min-width: 0;
}

<div
    class="emotion-3"
  >
    <div
      class="emotion-2"
    >
      <div
        class="emotion-1"
      >
        See also: 
        <a
          class="emotion-0"
          href="/foobar"
        >
          Foobar
        </a>
      </div>
      <div>
        Icon: lightbulb-on-outline
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`learn more renders correctly with multiple links 1`] = `
<DocumentFragment>
  .emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.emotion-4 {
  box-sizing: border-box;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-3 {
  box-sizing: border-box;
  min-width: 0;
}

<div
    class="emotion-5"
  >
    <div
      class="emotion-4"
    >
      <div
        class="emotion-3"
      >
        <div>
          See also:
        </div>
        <div>
          <a
            class="emotion-0"
            href="/foobar"
          >
            a
          </a>
        </div>
        <div>
          <a
            class="emotion-0"
            href="/foobar"
          >
            b
          </a>
        </div>
        <div>
          <a
            class="emotion-0"
            href="/foobar"
          >
            c
          </a>
        </div>
      </div>
      <div>
        Icon: lightbulb-on-outline
      </div>
    </div>
  </div>
</DocumentFragment>
`;
